### PR TITLE
Fix deadlines in DSProxy for Status, MultiPut, Patch requests, add UTs for deadlines

### PR DESCRIPTION
**Pull request by serbel324**
### Changelog entry 

Add UT for restore and index Get request deadlines

### Changelog category 

* unit tests

### Additional information

...

### CODE REVIEW

File: ydb/core/blobstorage/dsproxy/dsproxy_put.cpp [hunk 1]
```diff
@@ -793,6 +796,19 @@ class TBlobStorageGroupPutRequest : public TBlobStorageGroupRequestActor {
...
+        TInstant now = TActivationContext::Now();
+        while (!PutDeadlines.empty()) {
+            TInstant deadline = *PutDeadlines.begin();
+            PutDeadlines.erase(PutDeadlines.begin());
+            if (deadline > now) {
+                Schedule(deadline, new TKikimrEvents::TEvWakeup);
+                return;
+            }
+        }
+        Schedule(TDuration::MilliSeconds(DsPutWakeupMs), new TKikimrEvents::TEvWakeup);
```
Comment (alexvru) [inline ydb/core/blobstorage/dsproxy/dsproxy_put.cpp:809]: do we really need to wake up every 1 minute? to do what?

File: ydb/core/blobstorage/dsproxy/dsproxy_put.cpp [hunk 2]
```diff
@@ -721,12 +724,12 @@ class TBlobStorageGroupPutRequest : public TBlobStorageGroupRequestActor {
...
         TPutImpl::TPutResultVec putResults;
         for (size_t blobIdx = 0; blobIdx < PutImpl.Blobs.size(); ++blobIdx) {
-            if (!PutImpl.Blobs[blobIdx].Replied && now > PutImpl.Blobs[blobIdx].Deadline) {
+            if (!PutImpl.Blobs[blobIdx].Replied && now >= PutImpl.Blobs[blobIdx].Deadline) {
                 PutImpl.PrepareOneReply(NKikimrProto::DEADLINE, blobIdx, LogCtx, "Deadline timer hit", putResults);
             }
         }
         ReplyAndDieWithLastResponse(putResults);
-        Schedule(TDuration::MilliSeconds(DsPutWakeupMs), new TKikimrEvents::TEvWakeup);
+        ScheduleWakeup();
```
Comment (alexvru) [inline ydb/core/blobstorage/dsproxy/dsproxy_put.cpp:732]: there's no need to schedule anything if actor had passed away a line before

File: ydb/core/blobstorage/dsproxy/dsproxy_put.cpp [hunk 3]
```diff
@@ -793,6 +796,19 @@ class TBlobStorageGroupPutRequest : public TBlobStorageGroupRequestActor {
             << " State# " << PutImpl.DumpFullState());
     }
 
+    void ScheduleWakeup() {
+        TInstant now = TActivationContext::Now();
+        while (!PutDeadlines.empty()) {
+            TInstant deadline = *PutDeadlines.begin();
+            PutDeadlines.erase(PutDeadlines.begin());
+            if (deadline > now) {
+                Schedule(deadline, new TKikimrEvents::TEvWakeup);
```
Comment (alexvru) [inline ydb/core/blobstorage/dsproxy/dsproxy_put.cpp:805]: don't schedule for TInstant::Max()

**Review by alexvru — CHANGES_REQUESTED** [review]

File: ydb/core/blobstorage/dsproxy/dsproxy_put.cpp [hunk 4]
```diff
@@ -42,6 +42,10 @@ class TBlobStorageGroupPutRequest : public TBlobStorageGroupRequestActor {
     TErasureSplitContext ErasureSplitContext = TErasureSplitContext::Init(MaxBytesToSplitAtOnce);
     TBatchedVec<TStackVec<TRope, TypicalPartsInBlob>> PartSets;
 
+    static_assert(MaxBatchedPutRequests <= sizeof(ui64) * 8);
+    std::map<TInstant, ui64> PutDeadlineMasks;
```
Comment (va-kuznecov) [inline ydb/core/blobstorage/dsproxy/dsproxy_put.cpp:46]: https://en.cppreference.com/w/cpp/utility/bitset

```suggestion
    std::map<TInstant, std::bitset<MaxBatchedPutRequests>> PutDeadlineMasks;
```

Comment (va-kuznecov) [reply ydb/core/blobstorage/dsproxy/dsproxy_put.cpp:46]: please, don't add new asserts unless necessary

File: ydb/core/blobstorage/ut_blobstorage/deadlines.cpp [hunk 5]
```diff
@@ -102,95 +110,426 @@ Y_UNIT_TEST_SUITE(Deadlines) {
...
+            });
+        };
+
+        TestImpl<TEvBlobStorage::TEvVStatusResult, TEvBlobStorage::TEvStatusResult>(erasure, delay, timeout, send);
+    }
+
+    template<class Tag>
+    void Test(const TBlobStorageGroupType&) {}
+
+    struct TTagPut {};
```
Comment (va-kuznecov) [inline ydb/core/blobstorage/ut_blobstorage/deadlines.cpp:331]: What are these Tags supposed to be used for?

You can create functions with the same naming system as well and get rid of the third(!) request type list of in this file

**Review by alexvru — DISMISSED** [review]

File: ydb/core/blobstorage/ut_blobstorage/deadlines.cpp [hunk 6]
```diff
@@ -76,15 +80,19 @@ Y_UNIT_TEST_SUITE(Deadlines) {
 
         ui32 NodeCount;
         TBlobStorageGroupType Erasure;
-        std::shared_ptr<TEnvironmentSetup> Env;
+        std::shared_ptr<TEnvironmentSetup> Env = nullptr;
```
Comment (va-kuznecov) [inline ydb/core/blobstorage/ut_blobstorage/deadlines.cpp:83]:
```suggestion
        std::shared_ptr<TEnvironmentSetup> Env;
```

File: ydb/core/blobstorage/ut_blobstorage/deadlines.cpp [hunk 7]
```diff
@@ -76,15 +80,19 @@ Y_UNIT_TEST_SUITE(Deadlines) {
...
 
         ui32 GroupId;
         TActorId Edge;
         TDuration VDiskDelay;
-        std::shared_ptr<TVDiskDelayEmulator> VDiskDelayEmulator;
+        std::shared_ptr<TVDiskDelayEmulator> VDiskDelayEmulator = nullptr;
```
Comment (va-kuznecov) [inline ydb/core/blobstorage/ut_blobstorage/deadlines.cpp:88]:
```suggestion
        std::shared_ptr<TVDiskDelayEmulator> VDiskDelayEmulator;
```

**Review by va-kuznecov — CHANGES_REQUESTED** [review]

**Review by va-kuznecov — APPROVED** [review]

**Pull request merged**